sharding/indexHashedNodesCoordinator.go

```diff
@@ -451,49 +471,35 @@ func (ihgs *indexHashedNodesCoordinator) GetValidatorsIndexes(
 
 // EpochStartPrepare is called when an epoch start event is observed, but not yet confirmed/committed.
 // Some components may need to do some initialisation on this event
-func (ihgs *indexHashedNodesCoordinator) EpochStartPrepare(metaHeader data.HeaderHandler) {
-	randomness := metaHeader.GetPrevRandSeed()
-	newEpoch := metaHeader.GetEpoch()
-
-	ihgs.mutNodesConfig.RLock()
-	nodesConfig, ok := ihgs.nodesConfig[newEpoch-1]
-	ihgs.mutNodesConfig.RUnlock()
+func (ihgs *indexHashedNodesCoordinator) EpochStartPrepare(hdr data.HeaderHandler, body data.BodyHandler) {
+	randomness := hdr.GetPrevRandSeed()
+	newEpoch := hdr.GetEpoch()
 
-	if !ok {
-		log.Error("no configured epoch found")
+	allValidatorInfo, err := createValidatorInfoFromBody(body, ihgs.marshalizer, ihgs.numTotalEligible)
+	if err != nil {
+		log.Error("could not create validator info from body - do nothing on nodesCoordinator epochStartPrepare")
 		return
 	}
 
-	allValidators := make([]Validator, 0)
-
-	for _, shardValidators := range nodesConfig.eligibleMap {
-		allValidators = append(allValidators, shardValidators...)
-	}
-
-	for _, shardValidators := range nodesConfig.waitingMap {
-		allValidators = append(allValidators, shardValidators...)
+	newNodesConfig, err := ihgs.computeNodesConfigFromList(allValidatorInfo)
```

Comment:
If in the same epoch, I would check against current configuration for the same epoch

If I understand correctly the nodes list is the configuration from the previous epoch?
What about the rating?
And one other thing, how do validator metachain nodes verify the proposed configuration?

Reply:
Validator metachain nodes create the same PeerMiniBlocks. Peer Miniblocks are created by all of them - and checked whether the leader did the same thing correctly.

Reply:
So a peerminiblock is created by all metachain nodes - by exporting the data from the validator statistics trie and creating these ShardValidatorinfos. The validators will check if the hash of the miniblocks they created is the same that the leader created. In case of shard this function is called only for valid metablocks, so those created above the previous + all the signatures.